util/get-vars.py

```diff
@@ -2,112 +2,219 @@
+for d in ctl["cases"]:
+    lat = d["lat"]
+    lon = d["lon"]
+    nz = d.get("nz")
+    atm_fp = Path(d["atm"])
+    sfc_fp = Path(d["sfc"])
+    out_fp = Path(d["out"])
+    if not out_fp.is_absolute():
+        out_fp = HERE / out_fp
+    print(f"lat={lat}, lon={lon}, nz={nz}, atm={atm_fp}, sfc={sfc_fp}, out={out_fp}")
+
+    src = {}
+
+    src["atm"] = None
+    if "atm" in unique_src_ids:
+        ds = (
+            xr.open_dataset(atm_fp)
+            .drop_vars(["lon", "lat"])
+            .rename({"grid_xt": "lon", "grid_yt": "lat"})
+            .sel(lat=lat, lon=lon, method="nearest")
+            .squeeze()
+        )
+
+        # Switch to sfc first
+        pmid, p = ds["pfull"].values, ds["phalf"].values
+        assert pmid[0] < pmid[-1] and p[0] < p[-1]
+        ds = ds.isel(pfull=slice(None, None, -1), phalf=slice(None, None, -1))
+
+        # Compute height
+        assert "z" not in ds.variables
+        assert "zmid" not in ds.variables
+        zsfc = ds["hgtsfc"]
+        dz = ds["delz"]
+        if (dz < 0).all():
+            dz *= -1
+        ztop = (zsfc + dz).cumsum("pfull")
+        z = xr.DataArray(
+            data=np.concatenate(([zsfc.values], ztop.values)),
+            dims="phalf",
+            # coords={"phalf": ds.coords["phalf"]},
+        )
+        zmid = ((z + z.shift(phalf=1)) / 2).isel(phalf=slice(1, None)).swap_dims(phalf="pfull")
+        ds = ds.assign_coords(z=z, zmid=zmid)
+
+        # Select levels
+        if nz is not None:
+            ds = ds.isel(pfull=slice(None, nz), phalf=slice(None, nz + 1))
+
+        # Store
+        src["atm"] = ds
+
+    src["sfc"] = None
+    if "sfc" in unique_src_ids:
+        ds = (
+            xr.open_dataset(sfc_fp)
+            .drop_vars(["lon", "lat"])
+            .rename({"grid_xt": "lon", "grid_yt": "lat"})
+            .sel(lat=lat, lon=lon, method="nearest")
+            .squeeze()
+        )
+
+        # Switch to sfc first
+        pmid, p = ds["pfull"].values, ds["phalf"].values
+        assert pmid[0] < pmid[-1] and p[0] < p[-1]
+        ds = ds.isel(pfull=slice(None, None, -1), phalf=slice(None, None, -1))
+
+        # Select levels (for cloud fraction)
+        if nz is not None:
+            ds = ds.isel(pfull=slice(None, nz), phalf=slice(None, nz + 1))
+
+        # Compute soil moisture variables
+        #
+        # soilw{1..4}: volumetric (fraction) soil moisture at 10, 30, 60, 100 cm
+        # Noah predicts values at layer midpoints:
+        # - [0, 10)    5
+        # - [10, 40)   25
+        # - [40, 100)  70
+        # - [100, 200) 150
+        #
+        # GWETTOP: 0--5 cm avg?
+        # GWETROOT: 10--100 cm avg?
+        ds["gwettop"] = ds["soilw1"]  # FIXME: this is value at 5 cm, not 0--5 cm avg
+        ds["gwetroot"] = (ds["soilw2"] * 30 + ds["soilw3"] * 60) / 90
+
+        # Combine soil moisture into one array
+        ds["soilm"] = xr.concat([ds[f"soilw{i}"] for i in range(1, 5)], dim="soil")
```

Comment:
note we should make this flexible - was just in a meeting today for SFS and they are talking about expanding the number of soil levels (perhaps to 9)

Reply:
Sure I can change it now to search for "soilw" variables instead of hardcoding the numbers. But a soil dimension in the dataset would be better if they can do that.

Reply:
Addressed in 7f7208d